aten/src/ATen/native/Distributions.h

```diff
@@ -0,0 +1,215 @@
+#include "TH/THMath.h"
+#include "ATen/Half.h"
+#ifdef __CUDA_ARCH__
+#include <nvfunctional>
+#endif
+
+namespace {
+
+#ifdef __CUDA_ARCH__
+#define nvfunction_or_function nvstd::function
+#define deviceforcuda __device__
+#else
+#define nvfunction_or_function std::function
+#define deviceforcuda
+// we cannot use std::isnan directly due to some incompatibility of
+// gcc constexpr'ing and nvcc
+#define isnan std::isnan
+#endif
+
+template<typename scalar_t>
+struct BaseSampler {
+  nvfunction_or_function<scalar_t(void)> sampler;
+  deviceforcuda BaseSampler(nvfunction_or_function<scalar_t(void)> sampler): sampler(sampler) {}
+  deviceforcuda scalar_t sample() {
+    return sampler();
+  }
+};
+
+// The function `sample_gamma` is
+// is adapted from Numpy's distributions.c implementation.
+// It is MIT licensed, so here is the copyright:
+
+/* Copyright 2005 Robert Kern (robert.kern@gmail.com)
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the
+ * "Software"), to deal in the Software without restriction, including
+ * without limitation the rights to use, copy, modify, merge, publish,
+ * distribute, sublicense, and/or sell copies of the Software, and to
+ * permit persons to whom the Software is furnished to do so, subject to
+ * the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+ * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+
+template<typename scalar_t>
+deviceforcuda scalar_t sample_gamma(scalar_t alpha, BaseSampler<scalar_t>& standard_uniform, BaseSampler<scalar_t>& standard_normal) {
+  scalar_t scale = 1.0;
+
+  // Boost alpha for higher acceptance probability.
+  if (alpha < 1.0) {
+    scale *= std::pow(1 - standard_uniform.sample(), static_cast<scalar_t>(1.0) / alpha);
```

Comment:
Can you confirm that it is ok that the rngs in this function are called a data-dependent number of times (all three sampler invocations). IIRC @rachtsingh had to do some special handling with the Poisson samplers to keep random number generators synchronized.

Reply:
It would seem that the comment applies to the MTGP32 sampler.
Previous merged PRs switch to the Philox_4x32_10 generator. My understanding of the NVIDIA documentation is that it does not suffer from this restriction.
https://docs.nvidia.com/cuda/curand/device-api-overview.html#bit-generation-3

Reply:
Yes, this is correct. In `Distributions.cu` the function is invoked with a 'thread-safe' `standard_uniform` and `standard_normal` sampler implemented via a Philox state.

Reply:
mmmm, I think it's not ok, because philox offset is increased just by one for the next initialization https://github.com/pytorch/pytorch/blob/e089849b4a2f9437c27c435bb6ffab8dfebbb43a/aten/src/ATen/native/cuda/Distributions.cu#L23-L27, whereas it would generate a few numbers here, thus offset should be increased by no less than the number of randoms generated not to repeat parts of the sequence.

Reply:
I see, I misread the curand documentation and assumed that a (offset, subsequence) pair gave a unique sequence. This should cause a problem for `curand_poisson`, too, right? Since that internally calls `curand_uniform` which will implicitly increase the offset. 

I thought a bit about ways to fix this, and keeping state valid between multiple kernel launches is really difficult - can we instead pass a parameter which is 'maximum number of calls expected' (with high probability) into `next_philox_seed` and increment based on that? We seem to have a lot of wiggle room (2^64) and for each call to `next_philox_seed` it's a bit painful but possible to compute this upper bound (based on sampling algorithm).

Reply:
Yeah, it's either that, or actually preserving philox counters between kernel launches (which would make everything else more annoying, because you'd want then as few counters as possible => as few threadblocks as possible while still occupying the hardware => hardware-dependent launch code).

Reply:
Thank you @ngimel and @rachtsingh for your comments! They are invaluable.
So here is a theory I have been meaning to discuss for a long time: If we don't use a fixed offset increment but a uniformly random offset (with wrapping to the legal range) would that fix the issue?

Alternatively, could we increment sequence every time and not worry about the offset?

How do we resolve this? My idea would be to write a small test with uniform and applying the [Diehard](https://en.wikipedia.org/wiki/Diehard_tests) tests to see whether the quality is OK (relative to generating sequentially).
As @rachtsingh points out, that would also be in the `curand_poisson` and so would be somewhat orthogonal.

Reply:
The purist in me says no to everything ("anyone who considers arithmetical methods of producing random digits is, of course, in a state of sin"), but practically speaking using uniformly random offset should work, I think, if Diehard is happy.

Reply:
@rachtsingh it is usually faster and more reliable just to run the thing a few (million) times and count your successes, than to try to wrap your head around it.

Reply:
Thinking about it, randomly reseeding seems like a not so good idea. So I'd probably go with Rachit's idea of having a largish offset increase.

Reply:
- So to have some more safety margin, say 32 calls for a gamma sample as a probable upper bound. Do we want to do something for like that poisson as well?
- I think we would need to multiply this by round_up(alpha.numel() / number of threads) to capture we are sampling tensors. What is our assumption about the number of threads here?

Reply:
For all cards except Fermi it's a pretty safe bet that there'll be one element per thread https://github.com/pytorch/pytorch/blob/1848cad10802db9fa0aa066d9de195958120d863/aten/src/ATen/cuda/CUDAApplyUtils.cuh#L239-L254

Reply:
OK, so I need to increase offset by <random calls per sample> * <number of samples to fill tensor>, right?

Reply:
just `random_calls_per_sample`? Number of samples = number of threads, right? Each thread is already using it's own subsequence.